examples/chat/src/util/auth.rs

```diff
@@ -0,0 +1,67 @@
+use http::Extensions;
+use via::request::RequestHead;
+use via::{Middleware, raise, ws};
+
+use crate::chat::Chat;
+use crate::models::user::User;
+use crate::{Next, Request};
+
+pub trait Authenticate {
+    fn current_user(&self) -> via::Result<&User>;
+}
+
+pub struct Auth {
+    cookie: String,
+}
+
+#[derive(Clone)]
+struct Verify(User);
```

Comment:
This prevents malicious middlewares from accessing the `CurrentUser` stored in the request extensions.

Hyper does the same thing for websocket upgrades and should be suitable for any auth plugin someone may be considering writing.

I can't imagine any other option being more secure. The type visibility rules of Rust prevent this type from being spoofed. We believe the borrow checker when it gives us compiler errors related to memory safety so we should do the same and accept that this type cannot be named outside of this module.